pkg/steps/index_generator.go

```diff
@@ -0,0 +1,128 @@
+package steps
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	buildapi "github.com/openshift/api/build/v1"
+	"github.com/openshift/ci-tools/pkg/api"
+	"github.com/openshift/ci-tools/pkg/results"
+	imageclientset "github.com/openshift/client-go/image/clientset/versioned/typed/image/v1"
+	coreapi "k8s.io/api/core/v1"
+)
+
+type indexGeneratorStep struct {
+	config      api.IndexGeneratorStepConfiguration
+	resources   api.ResourceConfiguration
+	buildClient BuildClient
+	imageClient imageclientset.ImageStreamsGetter
+	istClient   imageclientset.ImageStreamTagsGetter
+	jobSpec     *api.JobSpec
+	artifactDir string
+	pullSecret  *coreapi.Secret
+}
+
+const IndexOutputDir = "/index-data"
+const IndexDockerfileName = "index.Dockerfile"
+
+func (s *indexGeneratorStep) Inputs() (api.InputDefinition, error) {
+	return nil, nil
+}
+
+func (s *indexGeneratorStep) Run(ctx context.Context) error {
+	return results.ForReason("building_index_generator").ForError(s.run(ctx))
+}
+
+func (s *indexGeneratorStep) run(ctx context.Context) error {
+	source := fmt.Sprintf("%s:%s", api.PipelineImageStream, api.PipelineImageStreamTagReferenceSource)
+	workingDir, err := getWorkingDir(s.istClient, source, s.jobSpec.Namespace())
+	if err != nil {
+		return fmt.Errorf("failed to get workingDir: %w", err)
+	}
+	dockerfile, err := s.indexGenDockerfile()
+	if err != nil {
+		return err
+	}
+	build := buildFromSource(
+		s.jobSpec, api.PipelineImageStreamTagReferenceSource, s.config.To,
+		buildapi.BuildSource{
+			Type:       buildapi.BuildSourceDockerfile,
+			Dockerfile: &dockerfile,
+			Images: []buildapi.ImageSource{
+				{
+					From: coreapi.ObjectReference{
+						Kind: "ImageStreamTag",
+						Name: source,
+					},
+					Paths: []buildapi.ImageSourcePath{{
+						SourcePath:     fmt.Sprintf("%s/.", workingDir),
+						DestinationDir: ".",
+					}},
+				},
+			},
+		},
+		"",
+		s.resources,
+		s.pullSecret,
+	)
+	return handleBuild(ctx, s.buildClient, build, s.artifactDir)
+}
+
+func (s *indexGeneratorStep) indexGenDockerfile() (string, error) {
```

Comment:
IIRC if this is a Build, we need to inject the pull credentials for the bundle into it. If w run it as a Pod and get the output Dockerfile as an artifact, we can mount the credentials as a secret.

Reply:
I found another way to do it that I believe is reasonable. Secrets can be mounted as build sources in openshift builds. We can use a multistage docker build to where the builder stage copies the pull secret to the correct location and runs the `opm` command and then use a scratch image as the output stage that just copies the database directory and dockerfile.

This should work with our registry and give us the data + dockerfile we need to build the actual index image.